Return control immediately to application on 401 errors other initialization failures

guilherme-santos opened the pull request.

guilherme-santos committed:
Fix issue #120

ashanbrown commented:
@guilherme-santos Thanks for your contribution.  My understanding is that you'd like to return control to your app immediately when the response is 401 (unauthorized due to a bad key) rather than waiting for the full initialization timeout.  Is that correct?  That sounds like a reasonable expectation.  Is it possible to implement this without changing the public interface?  There is a related change we are also considering which involves checking initialization success after we complete our initialization attempt and introducing a separate error response indicating that we failed to initialize immediately.  That might be relevant here. Thanks.

guilherme-santos commented:
Exactly this that I'd like. I'll try to do something without break the public API.

guilherme-santos committed:
Try to solve bug without break public api

guilherme-santos committed:
Fix tests

guilherme-santos commented:
@ashanbrown I fixed to not break the public api, but the return is not the real error instead will be only `Error initializing LaunchDarkly`, IMO we should refactor to return the proper error but as you're already logging I think is fine for while, can you review it?

ashanbrown commented on streaming.go:
```diff
@@ -196,23 +198,23 @@ func (sp *streamProcessor) subscribe(closeWhenReady chan<- struct{}) {
 			// Halt immediately if we've been closed already
 			select {
 			case <-sp.halt:
+				close(closeWhenReady)
```
I'm concerned that we might try to call `close` multiple times on this channel.  I think we can avoid this by using `sync.Once` to ensure this only happens once.  We could define:

```
var closer sync.Once
closeOnce := func() {
   closer.Do(func() {
      close(closeWhenReady)
   })
}
```
at the beginning of the method and then call `closeOnce()`

In retrospect, I won't be using a channel that works based on closing again as part of an interface but at this point the interface is out there.  ;)

guilherme-santos replied:
Actually, I thought the same, but in the end, I had the impression that will not happen. But for sure that way is the safer to go.

ashanbrown commented on ldclient.go:
```diff
@@ -167,12 +168,16 @@ func MakeCustomClient(sdkKey string, config Config, waitFor time.Duration) (*LDC
 		select {
 		case <-closeWhenReady:
+			if !client.updateProcessor.Initialized() {
+				return nil, ErrInitializationFailed
```
We still need to return the client here instead of `nil`.  It is our policy to always return the client, because the client is still expected to serve default values to the application, even if it fails to connect to the service.  We are definitely at fault for not making this clear in our test suite.

ashanbrown commented on ldclient.go:
```diff
@@ -167,12 +168,16 @@ func MakeCustomClient(sdkKey string, config Config, waitFor time.Duration) (*LDC
 			if waitFor > 0 {
 				config.Logger.Println("Timeout exceeded when initializing LaunchDarkly client.")
-				return &client, ErrInitializationTimeout
+				return nil, ErrInitializationTimeout
```
Again, let's return the client.

ashanbrown reviewed (changes requested):
Thank you for the fantastic updates and all the additional tests.  I have a few comments for changes that I hope you can make it and then it should be ready to merge.

guilherme-santos committed:
Always return client even in case of failure

guilherme-santos committed:
Ensure that channel is closed just once

guilherme-santos commented:
Changes made!

ashanbrown committed:
Clean up handling of ready notification

Avoid duplicate closes and use defer to ensure we always complete
initialization.

ashanbrown committed:
Remove duplicate

guilherme-santos commented:
@ashanbrown can you check again?

ashanbrown commented:
@guilherme-santos I created PR https://github.com/guilherme-santos/go-client/pull/1 against your original PR which has some fixes I think we need.  If you can merge that, I can test our our combined changes.  Thanks.

ashanbrown committed:
Stop polling when we get a 401

guilherme-santos committed:
Merge pull request #1 from launchdarkly/ashanbrown/clean-up-ready-notifications

Clean up handling of ready notification

guilherme-santos commented:
Hey @ashanbrown I think now we're ready to go! :)

ashanbrown commented:
Thank you!
